File: src/wire/hub.ts
/**
 * Hub mm-RPC wire types — request/response shapes for every endpoint
 * mm-cli calls on `meta-me.uk/api/mm`.
 *
 * Naming: `Hub<Feature><Action>{Req|Resp}`. The feature/action pair is
 * what `hubApi(feature, action, payload)` sends; the prefix matches.
 *
 * Single source of truth — referenced from every command, never inlined.
 * Used both as the TS-side type contract and as the catalogue the
 * `specs/go-port/01-wire.md` doc renders from.
 *
 * Convention: optional fields use `?`. Nullable-on-the-wire fields use
 * `T | null` (Google APIs returning explicit `null` rather than absent).
 */

// ─── Calendar ──────────────────────────────────────────────────────────

export type HubCalendarEvent = {
	id: string;
	summary: string;
	start: string;
	end: string;
	location: string | null;
	htmlLink: string | null;
	attendees: number;
	allDay: boolean;
};

export type HubCalendarListReq = {
	days?: number;
	q?: string;
	accountSlug?: string;
};
export type HubCalendarListResp = {
	events: HubCalendarEvent[];
	rangeDays: number;
	accountSlug: string | null;
};

export type HubCalendarCreateReq = {
	title: string;
	when: string;
	end?: string;
	location?: string;
	description?: string;
	attendees?: string[];
	sendUpdates?: 'all' | 'externalOnly' | 'none';
	accountSlug?: string;
};
export type HubCalendarCreateResp = {
	id: string;
	htmlLink: string | null;
	summary: string;
	start: string;
	end: string;
};

// ─── Tasks ─────────────────────────────────────────────────────────────

export type HubTasksTask = {
	id: string;
	title: string;
	notes?: string;
	due?: string;
	status: 'needsAction' | 'completed';
	webViewLink?: string;
};

export type HubTasksGroup = {
	listId: string;
	listTitle: string;
	tasks: HubTasksTask[];
};

export type HubTasksListReq = {
	days?: number;
	all?: boolean;
	accountSlug?: string;
};
export type HubTasksListResp = {
	groups: HubTasksGroup[];
	accountSlug: string | null;
};

export type HubTasksAddReq = {
	title: string;
	due?: string;
	listTitle?: string;
	listId?: string;
	notes?: string;
	accountSlug?: string;
};
export type HubTasksAddResp = {
	id: string;
	listId: string;
	listTitle: string;
};

export type HubTasksCompleteReq = {
	listId: string;
	taskId: string;
	accountSlug?: string;
};
export type HubTasksCompleteResp = { ok: true };

// ─── Drive ─────────────────────────────────────────────────────────────

export type HubDriveFile = {
	id: string;
	name: string;
	mimeType: string;
	modifiedTime?: string;
	webViewLink?: string;
};

export type HubDriveListReq = {
	q?: string;
	max?: number;
	accountSlug?: string;
};
export type HubDriveListResp = {
	files: HubDriveFile[];
	accountSlug: string | null;
};

export type HubDriveCreateDocReq = {
	name: string;
	content: string;
	sourceMime: 'text/markdown' | 'text/plain' | 'text/html';
	accountSlug?: string;
};
export type HubDriveCreateDocResp = {
	id: string;
	name: string;
	webViewLink: string | null;
};

export type HubDriveUpdateReq = {
	fileId: string;
	name?: string;
	addParents?: string[];
	removeParents?: string[];
	accountSlug?: string;
};
export type HubDriveUpdateResp = {
	id: string;
	name: string;
	mimeType: string;
	parents: string[];
	modifiedTime: string | null;
	webViewLink: string | null;
};

// ─── Email — platform outbound log (admin) ─────────────────────────────

export type HubEmailLogRow = {
	id: string;
	userId: string | null;
	toAddress: string;
	subject: string;
	template: string | null;
	status: string;
	triggeredBy: string | null;
	createdAt: string;
	sentAt: string | null;
	failedAt: string | null;
	error: string | null;
	parentId: string | null;
};

export type HubEmailListReq = {
	status?: string;
	template?: string;
	q?: string;
	limit?: number;
};
export type HubEmailListResp = {
	rows: HubEmailLogRow[];
	nextCursor: string | null;
};

export type HubEmailGetReq = { id: string };
export type HubEmailGetResp = HubEmailLogRow & {
	fromAddress: string;
	bodyHtml: string;
	bodyText: string;
	templateParams: Record<string, unknown> | null;
	messageId: string | null;
};

export type HubEmailCreateReq = {
	to: string;
	subject: string;
	html: string;
	text?: string;
	template?: string;
};
export type HubEmailCreateResp = { id: string };

export type HubEmailSendReq = { id: string };
export type HubEmailSendResp = {
	success: boolean;
	error?: string;
};

export type HubEmailResendReq = { id: string };
export type HubEmailResendResp = {
	newId: string;
	success: boolean;
	error?: string;
};

// ─── Email — Gmail inbox (via gws-gateway) ─────────────────────────────

export type HubInboxMessage = {
	id: string;
	threadId: string;
	subject: string;
	from: string;
	to: string;
	date: string;
	snippet: string;
	unread: boolean;
};

export type HubInboxSearchReq = {
	q?: string;
	maxResults?: number;
	accountSlug?: string;
};
export type HubInboxSearchResp = {
	messages: HubInboxMessage[];
	accountSlug: string | null;
};

export type HubInboxReadReq = { id: string };
export type HubInboxReadResp = HubInboxMessage & {
	cc: string;
	body: string;
	labels: string[];
};

// ─── Instance (cross-app — used by `mm chat nodes` to discover agents) ─

export type HubInstanceListReq = {
	slugs?: string[];
};
export type HubInstance = {
	id: string;
	appSlug: string;
	name: string;
	url: string | null;
	isOwner: boolean;
};
export type HubInstanceListResp = {
	instances: HubInstance[];
};

File: internal/wire/hub.go
// Package wire holds the request/response structs for every mm-cli HTTP
// surface. Mirrors src/wire/ in the TS source — same JSON shapes, same
// field names. The Go port either codegens these from the TS or hand-
// translates 1:1; this file is the hand-translated version.
package wire

// ─── Calendar ──────────────────────────────────────────────────────────

type HubCalendarEvent struct {
	ID        string  `json:"id"`
	Summary   string  `json:"summary"`
	Start     string  `json:"start"`
	End       string  `json:"end"`
	Location  *string `json:"location"`
	HTMLLink  *string `json:"htmlLink"`
	Attendees int     `json:"attendees"`
	AllDay    bool   `json:"allDay"`
}

type HubCalendarListReq struct {
	Days        *int    `json:"days,omitempty"`
	Q           string  `json:"q,omitempty"`
	AccountSlug string  `json:"accountSlug,omitempty"`
}

type HubCalendarListResp struct {
	Events      []HubCalendarEvent `json:"events"`
	RangeDays   int                `json:"rangeDays"`
	AccountSlug *string            `json:"accountSlug"`
}

type HubCalendarCreateReq struct {
	Title       string   `json:"title"`
	When        string   `json:"when"`
	End         string   `json:"end,omitempty"`
	Location    string   `json:"location,omitempty"`
	Description string   `json:"description,omitempty"`
	Attendees   []string `json:"attendees,omitempty"`
	SendUpdates string   `json:"sendUpdates,omitempty"` // "all"|"externalOnly"|"none"
	AccountSlug string   `json:"accountSlug,omitempty"`
}

type HubCalendarCreateResp struct {
	ID       string  `json:"id"`
	HTMLLink *string `json:"htmlLink"`
	Summary  string  `json:"summary"`
	Start    string  `json:"start"`
	End      string  `json:"end"`
}

// ─── Tasks ─────────────────────────────────────────────────────────────

type HubTasksTask struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Notes       string `json:"notes,omitempty"`
	Due         string `json:"due,omitempty"`
	Status      string `json:"status"` // "needsAction"|"completed"
	WebViewLink string `json:"webViewLink,omitempty"`
}

type HubTasksGroup struct {
	ListID    string         `json:"listId"`
	ListTitle string         `json:"listTitle"`
	Tasks     []HubTasksTask `json:"tasks"`
}

type HubTasksListReq struct {
	Days        *int   `json:"days,omitempty"`
	All         bool   `json:"all,omitempty"`
	AccountSlug string `json:"accountSlug,omitempty"`
}

type HubTasksListResp struct {
	Groups      []HubTasksGroup `json:"groups"`
	AccountSlug *string         `json:"accountSlug"`
}

type HubTasksAddReq struct {
	Title       string `json:"title"`
	Due         string `json:"due,omitempty"`
	ListTitle   string `json:"listTitle,omitempty"`
	ListID      string `json:"listId,omitempty"`
	Notes       string `json:"notes,omitempty"`
	AccountSlug string `json:"accountSlug,omitempty"`
}

type HubTasksAddResp struct {
	ID        string `json:"id"`
	ListID    string `json:"listId"`
	ListTitle string `json:"listTitle"`
}

type HubTasksCompleteReq struct {
	ListID      string `json:"listId"`
	TaskID      string `json:"taskId"`
	AccountSlug string `json:"accountSlug,omitempty"`
}

type HubTasksCompleteResp struct {
	OK bool `json:"ok"`
}

// ─── Drive ─────────────────────────────────────────────────────────────

type HubDriveFile struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	MimeType     string `json:"mimeType"`
	ModifiedTime string `json:"modifiedTime,omitempty"`
	WebViewLink  string `json:"webViewLink,omitempty"`
}

type HubDriveListReq struct {
	Q           string `json:"q,omitempty"`
	Max         *int   `json:"max,omitempty"`
	AccountSlug string `json:"accountSlug,omitempty"`
}

type HubDriveListResp struct {
	Files       []HubDriveFile `json:"files"`
	AccountSlug *string        `json:"accountSlug"`
}

type HubDriveCreateDocReq struct {
	Name        string `json:"name"`
	Content     string `json:"content"`
	SourceMime  string `json:"sourceMime"` // "text/markdown"|"text/plain"|"text/html"
	AccountSlug string `json:"accountSlug,omitempty"`
}

type HubDriveCreateDocResp struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	WebViewLink *string `json:"webViewLink"`
}

type HubDriveUpdateReq struct {
	FileID        string   `json:"fileId"`
	Name          string   `json:"name,omitempty"`
	AddParents    []string `json:"addParents,omitempty"`
	RemoveParents []string `json:"removeParents,omitempty"`
	AccountSlug   string   `json:"accountSlug,omitempty"`
}

type HubDriveUpdateResp struct {
	ID           string   `json:"id"`
	Name         string   `json:"name"`
	MimeType     string   `json:"mimeType"`
	Parents      []string `json:"parents"`
	ModifiedTime *string  `json:"modifiedTime"`
	WebViewLink  *string  `json:"webViewLink"`
}

// ─── Email — platform outbound log (admin) ─────────────────────────────

type HubEmailLogRow struct {
	ID          string  `json:"id"`
	UserID      *string `json:"userId"`
	ToAddress   string  `json:"toAddress"`
	Subject     string  `json:"subject"`
	Template    *string `json:"template"`
	Status      string  `json:"status"`
	TriggeredBy *string `json:"triggeredBy"`
	CreatedAt   string  `json:"createdAt"`
	SentAt      *string `json:"sentAt"`
	FailedAt    *string `json:"failedAt"`
	Error       *string `json:"error"`
	ParentID    *string `json:"parentId"`
}

type HubEmailListReq struct {
	Status   string `json:"status,omitempty"`
	Template string `json:"template,omitempty"`
	Q        string `json:"q,omitempty"`
	Limit    *int   `json:"limit,omitempty"`
}

type HubEmailListResp struct {
	Rows       []HubEmailLogRow `json:"rows"`
	NextCursor *string          `json:"nextCursor"`
}

type HubEmailGetReq struct {
	ID string `json:"id"`
}

type HubEmailGetResp struct {
	HubEmailLogRow
	FromAddress    string                 `json:"fromAddress"`
	BodyHTML       string                 `json:"bodyHtml"`
	BodyText       string                 `json:"bodyText"`
	TemplateParams map[string]interface{} `json:"templateParams"`
	MessageID      *string                `json:"messageId"`
}

type HubEmailCreateReq struct {
	To       string `json:"to"`
	Subject  string `json:"subject"`
	HTML     string `json:"html"`
	Text     string `json:"text,omitempty"`
	Template string `json:"template,omitempty"`
}

type HubEmailCreateResp struct {
	ID string `json:"id"`
}

type HubEmailSendReq struct {
	ID string `json:"id"`
}

type HubEmailSendResp struct {
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

type HubEmailResendReq struct {
	ID string `json:"id"`
}

type HubEmailResendResp struct {
	NewID   string `json:"newId"`
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

// ─── Email — Gmail inbox (via gws-gateway) ─────────────────────────────

type HubInboxMessage struct {
	ID       string `json:"id"`
	ThreadID string `json:"threadId"`
	Subject  string `json:"subject"`
	From     string `json:"from"`
	To       string `json:"to"`
	Date     string `json:"date"`
	Snippet  string `json:"snippet"`
	Unread   bool   `json:"unread"`
}

type HubInboxSearchReq struct {
	Q           string `json:"q,omitempty"`
	MaxResults  *int   `json:"maxResults,omitempty"`
	AccountSlug string `json:"accountSlug,omitempty"`
}

type HubInboxSearchResp struct {
	Messages    []HubInboxMessage `json:"messages"`
	AccountSlug *string           `json:"accountSlug"`
}

type HubInboxReadReq struct {
	ID string `json:"id"`
}

type HubInboxReadResp struct {
	HubInboxMessage
	CC     string   `json:"cc"`
	Body   string   `json:"body"`
	Labels []string `json:"labels"`
}

// ─── Instance discovery ────────────────────────────────────────────────

type HubInstanceListReq struct {
	Slugs []string `json:"slugs,omitempty"`
}

type HubInstance struct {
	ID      string  `json:"id"`
	AppSlug string  `json:"appSlug"`
	Name    string  `json:"name"`
	URL     *string `json:"url"`
	IsOwner bool    `json:"isOwner"`
}

type HubInstanceListResp struct {
	Instances []HubInstance `json:"instances"`
}

// ─── Feedback ──────────────────────────────────────────────────────────

type HubFeedbackSubmitReq struct {
	Message   string `json:"message"`
	AppSlug   string `json:"appSlug,omitempty"`
	URL       string `json:"url,omitempty"`
	UserAgent string `json:"userAgent,omitempty"`
}

type HubFeedbackSubmitResp struct {
	ID string `json:"id"`
}

type HubCaptureCreateReq struct {
	Text   string `json:"text"`
	Source string `json:"source,omitempty"`
}

type HubCaptureCreateResp struct {
	ID        string `json:"id"`
	Status    string `json:"status"`
	CreatedAt string `json:"createdAt"`
}

type HubCaptureListReq struct {
	Limit int `json:"limit,omitempty"`
}

type HubCaptureProposal struct {
	Target       HubCaptureClassifyTarget        `json:"target"`
	Payload      map[string]any               `json:"payload"`
	Confidence   float64                      `json:"confidence"`
	Rationale    string                       `json:"rationale"`
	Alternatives []HubCaptureClassifyAlternative `json:"alternatives"`
	Model        string                       `json:"model"`
	ClassifiedAt string                       `json:"classifiedAt"`
}

type HubCaptureListItem struct {
	ID        string           `json:"id"`
	Text      string           `json:"text"`
	Source    string           `json:"source"`
	Status    string           `json:"status"`
	CreatedAt string           `json:"createdAt"`
	Proposal  *HubCaptureProposal `json:"proposal,omitempty"`
}

type HubCaptureListResp struct {
	Items []HubCaptureListItem `json:"items"`
}

type HubCaptureClassifyReq struct {
	CaptureID string `json:"captureId"`
	Retry     bool   `json:"retry,omitempty"`
	Hint      string `json:"hint,omitempty"`
}

type HubCaptureClassifyTarget struct {
	Feature string `json:"feature"`
	Action  string `json:"action"`
}

type HubCaptureClassifyAlternative struct {
	Feature    string  `json:"feature"`
	Action     string  `json:"action"`
	Confidence float64 `json:"confidence"`
	Rationale  string  `json:"rationale"`
}

type HubCaptureClassifyResp struct {
	CaptureID    string                       `json:"captureId"`
	Text         string                       `json:"text"`
	Target       HubCaptureClassifyTarget        `json:"target"`
	Payload      map[string]any               `json:"payload"`
	Confidence   float64                      `json:"confidence"`
	Rationale    string                       `json:"rationale"`
	Alternatives []HubCaptureClassifyAlternative `json:"alternatives"`
	Model        string                       `json:"model"`
	LatencyMs    int                          `json:"latencyMs"`
}

type HubCaptureApproveReq struct {
	CaptureID       string         `json:"captureId"`
	Override        bool           `json:"override,omitempty"`
	PayloadOverride map[string]any `json:"payloadOverride,omitempty"`
}

type HubCaptureApproveOutcome struct {
	Kind    string `json:"kind"`
	Data    any    `json:"data,omitempty"`
	Code    string `json:"code,omitempty"`
	Message string `json:"message,omitempty"`
	Target  string `json:"target,omitempty"`
}

type HubCaptureApproveResp struct {
	CaptureID string                `json:"captureId"`
	Status    string                `json:"status"`
	Target    HubCaptureClassifyTarget `json:"target"`
	Outcome   HubCaptureApproveOutcome `json:"outcome"`
}

type HubCaptureDiscardReq struct {
	CaptureID string `json:"captureId"`
}

type HubCaptureDiscardResp struct {
	CaptureID string `json:"captureId"`
	Status    string `json:"status"`
}

// ─── Audit / Hermes runs ────────────────────────────────────────────────

type HubAuditListReq struct {
	Limit int    `json:"limit,omitempty"`
	Mode  string `json:"mode,omitempty"` // "run" for Hermes runs
}

type HubAuditRunSummary struct {
	RunID             string   `json:"runId"`
	CreatedAt         string   `json:"createdAt"`
	Status            string   `json:"status"`
	AppSlugs          []string `json:"appSlugs"`
	TotalGapsFound    int      `json:"totalGapsFound"`
	TotalGapsFixed    int      `json:"totalGapsFixed"`
	TotalFilesChecked int      `json:"totalFilesChecked"`
	Lookback          string   `json:"lookback"`
	Mode              string   `json:"mode"`
	Report            string   `json:"report"`
}

type HubAuditListResp struct {
	Runs []HubAuditRunSummary `json:"runs"`
}

type HubAuditShowReq struct {
	RunID string `json:"runId"`
}

type HubAuditRunDetail struct {
	ID           string `json:"id"`
	AppSlug      string `json:"appSlug"`
	CreatedAt    string `json:"createdAt"`
	Lookback     string `json:"lookback"`
	Mode         string `json:"mode"`
	FilesChecked int    `json:"filesChecked"`
	GapsFound    int    `json:"gapsFound"`
	GapsFixed    int    `json:"gapsFixed"`
	HighCount    int    `json:"highCount"`
	MediumCount  int    `json:"mediumCount"`
	LowCount     int    `json:"lowCount"`
	Report       string `json:"report"`
	Status       string `json:"status"`
}

type HubAuditShowResp struct {
	RunID     string              `json:"runId"`
	CreatedAt string              `json:"createdAt"`
	Status    string              `json:"status"`
	Rows      []HubAuditRunDetail `json:"rows"`
}

// ─── Hub conversations ─────────────────────────────────────────────────

type HubConversation struct {
	ID        string `json:"id"`
	Title     string `json:"title"`
	CreatedAt string `json:"createdAt"`
	UpdatedAt string `json:"updatedAt"`
	Pinned    bool   `json:"pinned"`
}

type HubConversationsListResp struct {
	Conversations []HubConversation `json:"conversations"`
}

type HubConversationMessage struct {
	ID        string  `json:"id"`
	Role      string  `json:"role"`
	Content   string  `json:"content"`
	CreatedAt string  `json:"createdAt"`
	ModelID   *string `json:"modelId"`
}

type HubMessagesListResp struct {
	Messages []HubConversationMessage `json:"messages"`
}

// ─── Envelope ──────────────────────────────────────────────────────────

// HubEnvelope is the top-level response shape for /api/mm and /api/v2.
// On success: { data: T }. On failure: { errors: [...] }.
type HubEnvelope[T any] struct {
	Data   T            `json:"data,omitempty"`
	Errors []HubErrItem `json:"errors,omitempty"`
}

type HubErrItem struct {
	Code   string `json:"code,omitempty"`
	Title  string `json:"title,omitempty"`
	Detail string `json:"detail,omitempty"`
	Status string `json:"status,omitempty"`
}
